Clean backoff record earlier

linxiulei opened the pull request:
#### What type of PR is this?
/kind bug


#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:

Fixes #132042

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fix a bug causing unexpected delay of creating pods for newly created jobs
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

mimowo commented on pkg/controller/job/job_controller.go:
```diff
@@ -2122,3 +2127,34 @@ func managedByExternalController(jobObj *batch.Job) *string {
+
+	jobObj, ok := obj.(*batch.Job)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
```
This code looks duplicated from `deleteJob`. Please extract to a helper function like `asJobObj(obj interface{}) (*batch.Job, error)`

mimowo commented on pkg/controller/job/job_controller.go:
```diff
@@ -2122,3 +2127,34 @@ func managedByExternalController(jobObj *batch.Job) *string {
+	case 2:
+		return parts[0], parts[1], nil
+	case 3:
+		return strings.Join(parts[:2], "/"), parts[2], nil
```
I would suggest avoid splitting and joining, maybe just:

```golang

func splitKeyUID(keyUID string) (string, string, error) {
	lastSlashIndex := strings.LastIndex(keyUID, "/")
	if lastSlashIndex != -1 {
		return keyUID[:lastSlashIndex], keyUID[lastSlashIndex+1:], nil
	}
	return "", "", fmt.Errorf("unexpected keyUID format: %q", keyUID)
}
```

mimowo commented on pkg/controller/job/job_controller.go:
```diff
@@ -796,13 +797,17 @@ func (jm *Controller) getPodsForJob(ctx context.Context, j *batch.Job) ([]*v1.Po
 // it did not expect to see any more of its pods created or deleted. This function is not meant to be invoked
 // concurrently with the same key.
-func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
+func (jm *Controller) syncJob(ctx context.Context, keyUID string) (rErr error) {
```
Isn't this always just "key"? Can you provide reference where we actually enqueue keyUID? Curious because I don't see any change when invoking `jm.queue.AddAfter` actually

mimowo commented:
@linxiulei I left some comments. Overall I think this approach could work, but I have some hesitations about corner cases races.

I'm wondering if another approach is simpler, like clearing the backoffStore in case the Job is newly created, based on empty Status (this way we can tell this is the first reconcile). wdyt?

linxiulei commented:
> @linxiulei I left some comments. Overall I think this approach could work, but I have some hesitations about corner cases races.

Thanks for the comments. Most comments are outdated after adopting your solution.

> 
> I'm wondering if another approach is simpler, like clearing the backoffStore in case the Job is newly created, based on empty Status, for example detected by checking that status.startTime=nil (this way we can tell this is the first reconcile). wdyt?

I like this solution, which is much simpler and less error prone. Changed.

linxiulei commented:
FTR: I manually tested the PR with the following hack in the code to make the repro easier

```
 		AddFunc: func(obj interface{}) {
+			time.Sleep(time.Second)
 			jm.addJob(logger, obj)
 		},
 		UpdateFunc: func(oldObj, newObj interface{}) {
+			time.Sleep(time.Second)
 			jm.updateJob(logger, oldObj, newObj)
 		},
 		DeleteFunc: func(obj interface{}) {
+			time.Sleep(time.Second)
 			jm.deleteJob(logger, obj)
 		},
```

The current code would print `"enqueueing job" logger="job-controller" key="default/base-100-workers-1" delay="2m30.705544372s"` to suggest the bug is reproduced.

Applying this PR, it won't print above log.

mimowo commented on pkg/controller/job/job_controller.go:
```diff
@@ -2122,3 +2132,14 @@ func managedByExternalController(jobObj *batch.Job) *string {
 }
+
+func newlyCreated(job *batch.Job) bool {
+	structIterator := reflect.ValueOf(job.Status)
+	for i := 0; i < structIterator.NumField(); i++ {
+		val := structIterator.Field(i).Interface()
+		if !reflect.DeepEqual(val, reflect.Zero(structIterator.Field(i).Type()).Interface()) {
+			return false
+		}
+	}
+	return true
```
This is ok, but rather brute-force. I think it is enough to check is `status.startTime=nil && no conditions`. This should cover all cases, unless some users manipulate the status, but if they do, then technically they could remove the whole status as well.

mimowo replied:
cc @soltysh @atiratree if you folks have any suggestions about the best way of solving this.

linxiulei replied:
I don't think the processing order of events is guaranteed, job controller puts the events in a queue and there are multiple workers to dequeue tasks, so the event processing is concurrent.

So alternatives can be:

* add more checks, if feasible, to determine it's really "newly created"
* only call `removeBackoffRecord()` without touch the `jm.expectations`, so that at least we fix the bug.

WDYT?

linxiulei replied:
That makes sense! Changed as suggested.

atiratree replied:
+1 for having it inside the deleteJob handler

mimowo commented on pkg/controller/job/job_controller.go:
```diff
@@ -810,23 +810,33 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 		return err
 	}
 
+	if newlyCreated(sharedJob) {
```
Add a comment explaining why we do so. 

```
Make sure the local in-memory state is cleaned when handling a newly created Job.
This is needed in case we still have some in-memory state left by the previous Job
with the same key (name & namespace). We use selected status fields to determine
the Job is newly created.
```

linxiulei committed:
Clean backoff record earlier

Once received job deletion event, it cleans the backoff records for that
job before enqueueing this job so that we can avoid a race condition
that the syncJob() may incorrect use stale backoff records for a newly created
job with same key.

Co-authored-by: Michal Wozniak <michalwozniak@google.com>

atiratree commented on pkg/controller/job/job_controller.go:
```diff
@@ -527,6 +526,16 @@ func (jm *Controller) updateJob(logger klog.Logger, old, cur interface{}) {
+	err = jm.podBackoffStore.removeBackoffRecord(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("error removing backoff record %w", err))
+		return
```
We should ensure that the rest of the function always executes, even if these early returns are unlikely. Just moving it to the end should be enough.

linxiulei commented:
> @linxiulei I got a bit of time and wrote the test under #132146. PTAL. Locally it reproduces the issue pretty much every time, and shows the fix works. Feel free to cherry-pick onto your branch. Feel free to also make cleanups or adjustments.

Thanks a lot. 

Didn't know what way to better to pull it into this PR. It could be separate commits but I decided to squash them into one and added your name to it. Please let me know if you prefer a different way.

mimowo commented on test/integration/job/job_test.go:
```diff
@@ -2284,6 +2284,102 @@ func TestManagedBy_Reenabling(t *testing.T) {
+	t.Cleanup(closeFn)
+	ctx, cancel := startJobControllerAndWaitForCaches(t, restConfig)
+	t.Cleanup(cancel)
+	resetMetrics()
```
non-blocking nit: drop it

atiratree commented on test/integration/job/job_test.go:
```diff
@@ -2284,6 +2284,101 @@ func TestManagedBy_Reenabling(t *testing.T) {
+	}
+
+	for i := 0; i < len(jobObjs); i++ {
+		jobObj := jobObjs[i].DeepCopy()
```
Ok, since we do not mutate the job objects, it would be better to remove the DeepCopy for better readability.

mimowo commented on test/integration/job/job_test.go:
```diff
@@ -2284,6 +2284,100 @@ func TestManagedBy_Reenabling(t *testing.T) {
+	}
+
+	for i := 0; i < len(jobObjs); i++ {
+		jobObj := jobObjs[i].DeepCopy()
```
nit, please drop DeepCopy as suggested in the other thread https://github.com/kubernetes/kubernetes/pull/132109/commits/b715a81277a907739608a94590c7f48d71c1924d#r2132605560

mimowo commented on test/integration/job/job_test.go:
```diff
@@ -2284,6 +2284,100 @@ func TestManagedBy_Reenabling(t *testing.T) {
+	for i := 0; i < len(jobObjs); i++ {
+		// wait maks 5s for the Active=1. This assert verifies that the backoff
+		// delay is not applied to the replacement instance of the Job.
+		jobObj := jobObjs[i].DeepCopy()
```
also here the deep copy was just precaution measure while testing, please attempt to drop to simplify

atiratree commented on test/integration/job/job_test.go:
```diff
@@ -2284,6 +2284,101 @@ func TestManagedBy_Reenabling(t *testing.T) {
+	}
+
+	var jobObjs []batchv1.Job
+	// create multiple Jobs we are going to recreate to make the issue more likely.
```
I see. We can put more to make the queue busy.

But then with more there is a small inconsistency with the timeouts
first pod can wait up to 5s, second up to 10s, sixth up to 30s.

The 2*ForeverTestTimeout is 60s so it works out, but if we would change some of these variables we could get an overlap. 

Could we paralelize or just wait the first one for 5s and the rest for 1s or 0s?

mimowo replied:
Ok, tested locally more, with 2 jobs I got failure 8/10.

atiratree replied:
@linxiulei ^

atiratree replied:
But let's document it and write the max time is less than 2*ForeverTestTimeout.

mimowo replied:
When the test is passing (fix applied) then on my local machine it takes about 10s with 3 jobs, but around 14s with 6 jobs.

mimowo commented on test/integration/job/job_test.go:
```diff
@@ -2284,6 +2284,99 @@ func TestManagedBy_Reenabling(t *testing.T) {
+	}
+
+	var jobObjs []*batchv1.Job
+	// create multiple Jobs we are going to recreate to make the issue more likely.
```
```suggestion
	// We create multiple Jobs to make the repro more likely. In particular, we need
	// more Jobs than the number of Job controller workers to make it very unlikely
	// that syncJob executes (and cleans the in-memory state) before the corresponding
	// replacement Jobs are created.
```

atiratree commented on test/integration/job/job_test.go:
```diff
@@ -2284,6 +2284,99 @@ func TestManagedBy_Reenabling(t *testing.T) {
+		}
+		jobObjs[i] = jobObj
+	}
+
+	for _, jobObj := range jobObjs {
```
```suggestion
    // total timeout (3*5s) is less than 2*ForeverTestTimeout.
	for _, jobObj := range jobObjs {
```

atiratree commented:
/triage accepted
/priority important
/approve

leaving the final tag on @mimowo

mimowo commented:
/lgtm

Leaving approve to @atiratree 

I'm supportive to cherry-pick, because:
- the risk is minimal given the size of the change and the amount of testing it received in the process, the change only affects the in-memory state of the backoff delay
- the issue is affecting JobSet quite often as it satisfies all conditions for the issue to appear: creates many Job instances, and reacts to Job failures by re-creating under the same, indexed name. For large Jobs with many Pods this can result in waiting very long for the replacement Pods.
